jdaviz/configs/cubeviz/plugins/cube_listener.py

```diff
@@ -0,0 +1,150 @@
+    with open(os.devnull, "w") as devnull:
+        old_stderr = sys.stderr
+        sys.stderr = devnull
+        try:
+            yield
+        finally:
+            sys.stderr = old_stderr
+
+
+def audify_spectrum(spec, duration, overlap=0.05, system='mono', srate=44100, fmin=40, fmax=1300,
+                    eln=False):
+    notes = [["A2"]]
+    score = Score(notes, duration)
+    # set up spectralizer generator
+    generator = Spectralizer(samprate=srate)
+
+    # Lets pick the mapping frequency range for the spectrum...
+    generator.modify_preset({'min_freq': fmin, 'max_freq': fmax,
+                             'fit_spec_multiples': False,
+                             'interpolation_type': 'preserve_power',
+                             'equal_loudness_normalisation': eln})
+
+    data = {'spectrum': [spec], 'pitch': [1]}
+
+    # again, use maximal range for the mapped parameters
+    lims = {'spectrum': ('0', '100')}
+
+    # set up source
+    sources = Events(data.keys())
+    sources.fromdict(data)
+    sources.apply_mapping_functions(map_lims=lims)
+
+    # render and play sonification!
+    soni = Sonification(score, sources, generator, system, samprate=srate)
+    soni.render()
+    soni._make_seamless(overlap)
+
+    return soni.loop_channels['0'].values
+
+
+class CubeListenerData:
+    def __init__(self, cube, wlens, samplerate=44100, duration=1, overlap=0.05, buffsize=1024,
+                 bdepth=16, wl_bounds=None, wl_unit=None, audfrqmin=50, audfrqmax=1500,
+                 eln=False, vol=None):
+        self.siglen = int(samplerate*(duration-overlap))
+        self.cube = cube
+        self.dur = duration
+        self.bdepth = bdepth
+        self.srate = samplerate
+        self.maxval = pow(2, bdepth-1) - 1
+        self.fadedx = 0
+
+        if vol is None:
+            self.atten_level = 1
+        else:
+            self.atten_level = int(np.clip((vol/100)**2, MINVOL, 1))
+
+        self.wl_bounds = wl_bounds
+        self.wl_unit = wl_unit
+        self.wlens = wlens
+
+        # control fades
+        fade = np.linspace(0, 1, buffsize+1)
+        self.ifade = fade[:-1]
+        self.ofade = fade[::-1][:-1]
+
+        # mapping frequency limits in Hz
+        self.audfrqmin = audfrqmin
+        self.audfrqmax = audfrqmax
+
+        # do we normalise for equal loudness?
+        self.eln = eln
+
+        self.idx1 = 0
+        self.idx2 = 0
+        self.cbuff = False
+        self.cursig = np.zeros(self.siglen, dtype='int16')
+        self.newsig = np.zeros(self.siglen, dtype='int16')
+
+        if self.cursig.nbytes * pow(1024, -3) > 2:
+            raise Exception("Cube projected to be > 2Gb!")
+
+        self.sigcube = np.zeros((*self.cube.shape[:2], self.siglen), dtype='int16')
+
+    def set_wl_bounds(self, w1, w2):
+        """
+        set the wavelength bounds for indexing spectra
+        """
+        wsrt = np.sort([w1, w2])
+        self.wl_bounds = tuple(wsrt)
+
+    def audify_cube(self):
```

Comment:
Given that `audify_spectrum` is outside the class as an independent function, and this pretty much just uses `self.cube` from `self`, I wonder if this could/should be pulled out into a function that takes `cube` as input and returns what's needed to the class 🤔. Probably not important, it just seemed a little wonky to have `audify_spectrum` outside this class and this inside. Maybe it's just that `audify_spectrum` would be better off in the sonify plugin file.

Reply:
Yeah I think that is what #3330 is hoping to fix by moving some/all of the `cube_listener.py` code into the sonify data plugin.